File: reach_definition/Write_Reach_Shp.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

	%SCRIPT FUNCTION:
        %Outputs reach locations and attributes in shapefile format at the
        %Pfafstetter level 2 basin scale.

    %INPUTS
        %nc_file -- SWORD NetCDF file.
        %outfilepath -- Path for shapefile to be written.
		%filename -- Name of shapefile.
		%version -- Version of SWORD.

    %OUTPUTS
        %SWORD Reach Shapefile -- Point shapefile containing locations and attributes
            %for the reaches (~ 10 km length) at the level 2 Pfafstetter basin scale.


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%clear variables.
clear

%set working directory.
cd C:/Users/ealtenau/Documents/Research/SWAG/For_Server/scripts/reach_definition/

%input/output files.
nc_file = '../../outputs/Reaches_Nodes/netcdf/eu_sword_v12.nc';
outfilepath='../../outputs/Reaches_Nodes/shp/EU/';
filename='eu_sword_reaches_hb';
version = '_v12';

%assigning variables to vectors.
cl_ind_all = ncread(nc_file,'/centerlines/cl_id');
reach_id_all = ncread(nc_file,'/centerlines/reach_id');
x_all = ncread(nc_file,'/centerlines/x');
y_all = ncread(nc_file,'/centerlines/y');
rch_x_all = ncread(nc_file,'/reaches/x');
rch_y_all = ncread(nc_file,'/reaches/y');
rch_id_all = ncread(nc_file,'/reaches/reach_id');
reach_len_all = ncread(nc_file,'/reaches/reach_length');
n_nodes_all = ncread(nc_file,'/reaches/n_nodes');
wse_all = ncread(nc_file,'/reaches/wse');
wse_var_all = ncread(nc_file,'/reaches/wse_var');
width_all = ncread(nc_file,'/reaches/width');
wth_var_all = ncread(nc_file,'/reaches/width_var');
facc_all = ncread(nc_file,'/reaches/facc');
n_chan_max_all = ncread(nc_file,'/reaches/n_chan_max');
n_chan_mod_all = ncread(nc_file,'/reaches/n_chan_mod');
grod_id_all = ncread(nc_file,'/reaches/obstr_type');
slope_all = ncread(nc_file,'/reaches/slope');
dist_out_all = ncread(nc_file,'/reaches/dist_out');
n_rch_up_all = ncread(nc_file,'/reaches/n_rch_up');
n_rch_down_all = ncread(nc_file,'/reaches/n_rch_down');
lakeflag_all = ncread(nc_file,'/reaches/lakeflag');
%lake_id_all = ncread(nc_file,'/reaches/lake_id');
rch_id_up_all = ncread(nc_file,'/reaches/rch_id_up');
rch_id_dn_all = ncread(nc_file,'/reaches/rch_id_dn');
grod_fid_all = ncread(nc_file,'/reaches/grod_id');
hfalls_fid_all = ncread(nc_file,'/reaches/hfalls_id');
num_obs_all = ncread(nc_file,'/reaches/swot_obs');
orbits_all = ncread(nc_file,'/reaches/swot_orbits');
rnames_all = h5read(nc_file,'/reaches/river_name');
max_wth_all = ncread(nc_file,'/reaches/max_width');

mod_ids_cl = mod(reach_id_all(:,1), 1000000000);
level2_cl = (reach_id_all(:,1)-mod_ids_cl)./1000000000;
mod_ids_rch = mod(rch_id_all, 1000000000);
level2_rch = (rch_id_all-mod_ids_rch)./1000000000;
uniq_level2 = unique(level2_cl);
%loop through every level 2 basin and output shapefiles.
for ind=1:length(uniq_level2)

    tic
    disp(['Starting Basin: ' num2str(uniq_level2(ind))])

    l2_cl = find(level2_cl == uniq_level2(ind));
    l2_rch = find(level2_rch == uniq_level2(ind));

    cl_ind = cl_ind_all(l2_cl);
    reach_id = reach_id_all(l2_cl,:);
    x = x_all(l2_cl);
    y = y_all(l2_cl);
    rch_x = rch_x_all(l2_rch);
    rch_y = rch_y_all(l2_rch);
    rch_id = rch_id_all(l2_rch);
    reach_len = reach_len_all(l2_rch);
    n_nodes = n_nodes_all(l2_rch);
    wse = wse_all(l2_rch);
    wse_var = wse_var_all(l2_rch);
    width = width_all(l2_rch);
    wth_var = wth_var_all(l2_rch);
    facc = facc_all(l2_rch);
    n_chan_max = n_chan_max_all(l2_rch);
    n_chan_mod = n_chan_mod_all(l2_rch);
    grod_id = grod_id_all(l2_rch);
    grod_fid = grod_fid_all(l2_rch);
    hfalls_fid = hfalls_fid_all(l2_rch);
    slope = slope_all(l2_rch);
    dist_out = dist_out_all(l2_rch);
    n_rch_up = n_rch_up_all(l2_rch);
    n_rch_down = n_rch_down_all(l2_rch);
    %lake_id = lake_id_all(l2_rch);
    lakeflag = lakeflag_all(l2_rch);
    rch_id_up = rch_id_up_all(l2_rch,:);
    rch_id_dn = rch_id_dn_all(l2_rch,:);
    num_obs = num_obs_all(l2_rch);
    orbits = orbits_all(l2_rch,:);
    rnames = rnames_all(l2_rch);
	max_wth = max_wth_all(l2_rch);
	
    %finding unique reaches to loop through
    reach_id_first=unique(reach_id(:,1));
    reach_id_first=reach_id_first(reach_id_first>0);
    
    %counting number of centerline points array for each reach.
    ct_pts = [];
    for rch=1:length(reach_id_first)
        rch_pts=find(reach_id(:,1)==reach_id_first(rch));
        check = length(rch_pts);
        ct_pts(rch) = check;
    end
    
    %finding where reaches only have one centerline point. 
    ct_pts = transpose(ct_pts);
    keep = find(ct_pts(:,1) > 1);
    
    %deleting reaches with only one point. 
    rch_id = rch_id(keep);
    rch_x = rch_x(keep);
    rch_y = rch_y(keep);
    reach_len = reach_len(keep);
    n_nodes = n_nodes(keep);
    wse = wse(keep);
    wse_var = wse_var(keep);
    width = width(keep);
    wth_var = wth_var(keep);
    facc = facc(keep);
    n_chan_max = n_chan_max(keep);
    n_chan_mod = n_chan_mod(keep);
    grod_id = grod_id(keep);
    grod_fid = grod_fid(keep);
    hfalls_fid = hfalls_fid(keep);
    slope = slope(keep);
    dist_out = dist_out(keep);
    n_rch_up = n_rch_up(keep);
    n_rch_down = n_rch_down(keep);
    %lake_id = lake_id(keep);
    lakeflag = lakeflag(keep);
    rch_id_up = rch_id_up(keep,:);
    rch_id_dn = rch_id_dn(keep,:);
    num_obs = num_obs(keep);
    orbits = orbits(keep,:);
    rnames = rnames(keep);
	max_wth = max_wth(keep);
	
    %finding unique reaches to loop through
    reach_id_un=unique(rch_id);
    reach_id_un=reach_id_un(reach_id_un>0);
    %looping through each reach and creating polyline.
    centerline = struct;
    for ct=1:length(reach_id_un) %test index -> ct=5381, 1358, 1344
        in_reach=find(reach_id(:,1)==reach_id_un(ct)); %position in the vector of points that belong to a reach
        var_ct = find(rch_id(:,1)==reach_id_un(ct)); %subset for variables
        
        %cl_ind when sorted gives you the rigth order of points, so I'm sorting them and using that order to sort the lon and lat vectors
        [~,sort_ind]=sort(cl_ind(in_reach));
        Lon=x(in_reach);
        Lat=y(in_reach);
        centerline(ct).Geometry='PolyLine';
        centerline(ct).BoundingBox=[[min(Lon) min(Lat)]; [max(Lon) max(Lat)]];
        centerline(ct).Lon=Lon(sort_ind);%save in the vector that will be used in the shapefiles
        centerline(ct).Lat=Lat(sort_ind);
        in_reach_up_dwn=[];
        for ct2=2:4
            %see if there are points that belong to this reach in other
            %positions of reach_id
            in_reach_up_dwn=[in_reach_up_dwn;find(reach_id(:,ct2)==reach_id_un(ct))];
        end
        if ~isempty(in_reach_up_dwn)
            for ct3=1:length(in_reach_up_dwn)
                %check if the point is closer to the first or last point in the
                %reach
                x_pt=x(in_reach_up_dwn(ct3));
                y_pt=y(in_reach_up_dwn(ct3));
                %distance to first point
                d1=deg2km(distance(centerline(ct).Lat(1), centerline(ct).Lon(1), y_pt, x_pt))*1000;
                d2=deg2km(distance(centerline(ct).Lat(end), centerline(ct).Lon(end), y_pt, x_pt))*1000;
                %don't connect if minimum distance greater than 200 meters. This is needed if there is an error with an identified neighbor in the database.
                if min(d1, d2) > 200
                    continue
                end
                if d1<d2
                    centerline(ct).Lon=[x_pt;centerline(ct).Lon];
                    centerline(ct).Lat=[y_pt;centerline(ct).Lat];
                end
                if d1>d2
                    centerline(ct).Lon=[centerline(ct).Lon;x_pt];
                    centerline(ct).Lat=[centerline(ct).Lat;y_pt];
                end
            end
        end

        r=reach_id_un(ct);
        centerline(ct).x=rch_x(var_ct);
        centerline(ct).y=rch_y(var_ct);
        centerline(ct).reach_id=num2str(r);
        centerline(ct).reach_len=reach_len(var_ct);
        centerline(ct).n_nodes=n_nodes(var_ct);
        centerline(ct).wse=wse(var_ct);
        centerline(ct).wse_var=wse_var(var_ct);
        centerline(ct).width=width(var_ct);
        centerline(ct).width_var=wth_var(var_ct);
        centerline(ct).facc=facc(var_ct);
        centerline(ct).n_chan_max=n_chan_max(var_ct);
        centerline(ct).n_chan_mod=n_chan_mod(var_ct);
        centerline(ct).obstr_type=grod_id(var_ct);
        centerline(ct).grod_id=grod_fid(var_ct);
        centerline(ct).hfalls_id=hfalls_fid(var_ct);
		centerline(ct).slope=slope(var_ct);
        centerline(ct).dist_out=dist_out(var_ct);
        centerline(ct).lakeflag=lakeflag(var_ct);
		centerline(ct).max_width=max_wth(var_ct);
        centerline(ct).n_rch_up=n_rch_up(var_ct);
        centerline(ct).n_rch_dn=n_rch_down(var_ct);
        centerline(ct).rch_id_up=num2str(rch_id_up(var_ct,find(rch_id_up(var_ct,:) > 0)));
        centerline(ct).rch_id_dn=num2str(rch_id_dn(var_ct,find(rch_id_dn(var_ct,:) > 0)));
        centerline(ct).swot_orbit=num2str(orbits(var_ct,find(orbits(var_ct,:) > 0)));
        centerline(ct).swot_obs=num_obs(var_ct);
        centerline(ct).type=double(mod(r,10));
        %centerline(ct).lake_id=num2str(lake_id(var_ct));
		centerline(ct).river_name=rnames(var_ct);
        %disp(ct) %print loop index
    end

    %write shapefiles
    outfilename=[outfilepath filename num2str(uniq_level2(ind)) version];
    shapewrite(centerline,outfilename)
    %write projection information
    string='GEOGCS["GCS_WGS_1984",DATUM["D_WGS_1984",SPHEROID["WGS_1984",6378137.0,298.257223563]],PRIMEM["Greenwich",0.0],UNIT["Degree",0.0174532925199433]]';
    fileID = fopen([outfilename '.prj'],'w');
    fprintf(fileID,string);
    fclose(fileID);

    toc
end

File: reach_definition/Write_Reach_Shp.py
import numpy as np
import netCDF4 as nc
import geopandas as gp
import geopy.distance
from shapely.geometry import LineString, Point
from geopandas import GeoSeries
import pandas as pd
import time

#############################################################################################

def define_geometry(unq_rch, reach_id, cl_x, cl_y):
    geom = []
    rm_ind = []
    for ind in list(range(len(unq_rch))):
        # print(ind)
        in_rch = np.where(reach_id[0,:] == unq_rch[ind])[0]
        sort_ind = in_rch[np.argsort(cl_id[in_rch])]
        x_coords = cl_x[sort_ind]
        y_coords = cl_y[sort_ind]

        #appending neighboring reach endpoints to coordinates
        in_rch_up_dn = []
        for ngh in list(range(1,4)):
            in_rch_up_dn.append(np.where(reach_id[ngh,:]==unq_rch[ind])[0])

        in_rch_up_dn = np.unique([j for sub in in_rch_up_dn for j in sub])
        if len(in_rch_up_dn) > 0:
            for ct in list(range(len(in_rch_up_dn))):
                x_pt = cl_x[in_rch_up_dn[ct]]
                y_pt = cl_y[in_rch_up_dn[ct]]
                
                if x_pt < 0 and np.min([cl_x[sort_ind[0]], cl_x[sort_ind[-1]]]) > 0:
                    print(unq_rch[ind])
                    continue
                
                else:
                    #distance to first and last point. 
                    coords_1 = (y_pt, x_pt)
                    coords_2 = (cl_y[sort_ind[0]], cl_x[sort_ind[0]])
                    coords_3 = (cl_y[sort_ind[-1]], cl_x[sort_ind[-1]])
                    d1 = geopy.distance.geodesic(coords_1, coords_2).m
                    d2 = geopy.distance.geodesic(coords_1, coords_3).m
                    #if minimum distance is greater than 200 m then don't attach. 
                    if np.min([d1,d2]) > 200:
                        continue
                    else:
                        if d1 < d2:
                            x_coords = np.insert(x_coords, 0, x_pt, axis=0)
                            y_coords = np.insert(y_coords, 0, y_pt, axis=0)
                        if d1 > d2: 
                            x_coords = np.insert(x_coords, len(x_coords), x_pt, axis=0)
                            y_coords = np.insert(y_coords, len(y_coords), y_pt, axis=0)

        pts = GeoSeries(map(Point, zip(x_coords, y_coords)))
        if len(pts) <= 1:
            rm_ind.append(ind)
            continue
        else:
            line = LineString(pts.tolist())
            geom.append(line) 

    return geom, rm_ind

#############################################################################################

#read in netcdf data. 
region = 'NA'
version = 'v14'
outdir = '/Users/ealteanau/Documents/SWORD_Dev/outputs/'
outpath = outdir+version+'/'
fn = outpath+'netcdf/'+region.lower()+'_sword_'+version+'.nc'
# fn = '/Users/ealteanau/Documents/SWORD_Dev/outputs/v14/netcdf/na_sword_v14_subset.nc'

data = nc.Dataset(fn)

#pull centerline level data. 
reach_id = data.groups['centerlines'].variables['reach_id'][:]
cl_id = data.groups['centerlines'].variables['cl_id'][:]
cl_x = data.groups['centerlines'].variables['x'][:]
cl_y = data.groups['centerlines'].variables['y'][:]
#identify unique reach ids. 
unq_rch = data.groups['reaches'].variables['reach_id'][:]

#reformat multi-dimensional variables
rch_type = [int(str(rch)[-1]) for rch in unq_rch]
rch_up = data.groups['reaches'].variables['rch_id_up'][:].T
rch_dn = data.groups['reaches'].variables['rch_id_dn'][:].T
swot_orbs = data.groups['reaches'].variables['swot_orbits'][:].T
rch_id_up = []; rch_id_dn = []; swot_orbits = []
for ind in list(range(len(rch_type))):
    rch_id_up.append(str(rch_up[ind,np.where(rch_up[ind,:] > 0)[0]])[1:-1])
    rch_id_dn.append(str(rch_dn[ind,np.where(rch_dn[ind,:] > 0)[0]])[1:-1])
    swot_orbits.append(str(swot_orbs[ind,np.where(swot_orbs[ind,:] > 0)[0]])[1:-1])

#create geometry for each reach. 
print('Creating Reach Geometry')
start = time.time()
geom, rm_ind = define_geometry(unq_rch, reach_id, cl_x, cl_y)
end = time.time()
print('Finished Reach Geometry in: '+str(np.round((end-start)/60,2))+' min')

#create initial GeoDataFrame.
reaches = gp.GeoDataFrame([
    data.groups['reaches'].variables['x'][:],
    data.groups['reaches'].variables['y'][:],
    data.groups['reaches'].variables['reach_id'][:],
    data.groups['reaches'].variables['reach_length'][:],
    data.groups['reaches'].variables['n_nodes'][:],
    data.groups['reaches'].variables['wse'][:],
    data.groups['reaches'].variables['wse_var'][:],
    data.groups['reaches'].variables['width'][:],
    data.groups['reaches'].variables['width_var'][:],
    data.groups['reaches'].variables['facc'][:],
    data.groups['reaches'].variables['n_chan_max'][:],
    data.groups['reaches'].variables['n_chan_mod'][:],
    data.groups['reaches'].variables['obstr_type'][:],
    data.groups['reaches'].variables['grod_id'][:],
    data.groups['reaches'].variables['hfalls_id'][:],
    data.groups['reaches'].variables['slope'][:],
    data.groups['reaches'].variables['dist_out'][:],
    data.groups['reaches'].variables['lakeflag'][:],
    data.groups['reaches'].variables['max_width'][:],
    data.groups['reaches'].variables['n_rch_up'][:],
    data.groups['reaches'].variables['n_rch_down'][:],
    rch_id_up,
    rch_id_dn,
    swot_orbits,
    data.groups['reaches'].variables['swot_obs'][:],
    rch_type,
    data.groups['reaches'].variables['river_name'][:],
    data.groups['reaches'].variables['edit_flag'][:],
    # data.groups['reaches'].variables['lake_id'][:],
]).T

#rename columns.
reaches.rename(
    columns={
        0:"x",
        1:"y",
        2:"reach_id",
        3:"reach_len",
        4:"n_nodes",
        5:"wse",
        6:"wse_var",
        7:"width",
        8:"width_var",
        9:"facc",
        10:"n_chan_max",
        11:"n_chan_mod",
        12:"obstr_type",
        13:"grod_id",
        14:"hfalls_id",
        15:"slope",
        16:"dist_out",
        17:"lakeflag",
        18:"max_width",
        19:"n_rch_up",
        20:"n_rch_dn",
        21:"rch_id_up",
        22:"rch_id_dn",
        23:"swot_orbit",
        24:"swot_obs",
        25:"type",
        26:"river_name",
        27:"edit_flag",
        # 28:"lake_id",
        },inplace=True)


#removing rows where reach was only one point.
reaches.drop(rm_ind, inplace=True)
#update data types
reaches = reaches.apply(pd.to_numeric, errors='ignore') # reaches.dtypes
#add geometry column and define crs. 
reaches['geometry'] = geom
reaches.set_geometry(col='geometry', inplace=True)
reaches = reaches.set_crs(4326, allow_override=True)

print('Writing GeoPackage File')
start = time.time()
#write geopackage (continental scale)
outgpkg = outpath + 'gpkg/' + region.lower() + '_sword_reaches_' + version + '.gpkg'
reaches.to_file(outgpkg, driver='GPKG', layer='reaches')
end = time.time()
print('Finished GPKG in: '+str(np.round((end-start)/60,2))+' min')

#write as shapefile per level2 basin.
print('Writing Shapefiles')
start = time.time()
level2 = [int(str(r)[0:2]) for r in reaches['reach_id']]
unq_l2 = np.unique(level2)
rch_cp = reaches.copy(); rch_cp['level2'] = level2
for lvl in list(range(len(unq_l2))):
    outshp = outpath + 'shp/' + region + '/' + region.lower() + "_sword_reaches_hb" + str(unq_l2[lvl]) + "_" + version + '.shp'
    subset = rch_cp[rch_cp['level2'] == unq_l2[lvl]]
    subset = subset.drop(columns=['level2'])
    subset.to_file(outshp)
    del(subset)
end = time.time()
print('Finished SHPs in: '+str(np.round((end-start)/60,2))+' min')
